platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemReporter.java

```diff
@@ -81,13 +84,40 @@ public DefaultProblemBuilder createProblemBuilder() {
         return new DefaultProblemBuilder(namespace);
     }
 
-    @Override
-    public void report(Problem problem) {
-        // Transform the problem with all registered transformers
+    private Problem transformProblem(Problem problem) {
         for (ProblemTransformer transformer : transformers) {
             problem = transformer.transform((InternalProblem) problem);
         }
+        return problem;
+    }
 
-        emitter.emit(problem);
+    /**
+     * Reports a problem.
+     * <p>
+     * The current build operation is used as the operation identifier.
+     * <p>
+     * If there is no current build operation, the problem is not reported.
+     *
+     * @param problem The problem to report.
+     */
+    @Override
+    public void report(Problem problem) {
+        OperationIdentifier id = CurrentBuildOperationRef.instance().getId();
+        report(problem, id);
+    }
+
+    /**
+     * Reports a problem with an explicit operation identifier.
+     * <p>
+     * If the operation identifier is null, the problem is not reported.
+     *
+     * @param problem The problem to report.
+     * @param id The operation identifier to associate with the problem.
+     */
+    @Override
+    public void report(Problem problem, @Nullable OperationIdentifier id) {
```

Comment:
❓ Javadoc aside, the method signature is a bit contradictory. IIUC this method should be called only if `id` is not null, otherwise it will fail silently (i.e. emit() is not called). For me the following API would be more clear to use 
```
report(Problem) // emits a problem with the current build operation
report(Problem, OperationIdentifier) // emits a problem with the specified operation id; throws exception if id is null
```
Then the clients needs to explicitly decide whether the report should happen (with a null check).

Reply:
@hegyibalint If you are merging, please prepare a follow-up issue/pr for this.

Reply:
https://github.com/gradle/gradle/issues/27416